plugins/de.cau.cs.kieler.klighd.lsp/src/de/cau/cs/kieler/klighd/lsp/KGraphIncrementalDiagramGenerator.xtend

```diff
@@ -0,0 +1,753 @@
+                newLabel.data += identifier
+                
+                // generate a new Label as if it would belong to the main model
+                val sKLabel = generateLabel(newLabel, false)
+                // All lines point towards the same original data. For matching, the index in the ID has to be taken
+                // into account as well to match it to its line.
+                textMapping.put(sKLabel.id, data)
+                
+                dataLabels += sKLabel
+            ]
+        } else if (data instanceof KContainerRendering) {
+            // KImages are container renderings themselves, so also look for their child renderings.
+            if (data instanceof KImage) {
+                imageData = ImageData.of(data)
+            }
+            
+            for (childData: data.children) {
+                findSpecialRenderings(childData)
+            }
+        } else if (data instanceof KRenderingLibrary) {
+            for (libraryEntry : data.renderings) {
+                if (libraryEntry instanceof KRendering) {
+                    findSpecialRenderings(libraryEntry)
+                }
+            }
+        }
+        if (!dataLabels.empty) {
+            modelLabels.addAll(dataLabels)
+        }
+        if (imageData !== null) {
+            images.add(imageData)
+        }
+    }
+    
+    /**
+     * Get method for the mapping from generated IDs to their corresponding {@link KGraphElement}s.
+     * 
+     * @return The mapping.
+     */
+    def getIdToKGraphElementMap() {
+        return idGen.idToElementMap
+    }
+
+    /**
+     * Returns a String describing the type of the {@link SModelElement}.
+     */
+    private static def String getTypeString(SModelElement element) {
+        switch element {
+            SNode: 'node'
+            SLabel: 'label'
+            SEdge: 'edge'
+            SPort: 'port'
+            default: throw new IllegalArgumentException("Unknown SModelElement type: "+ element?.class)
+        }
+    }
+
+    /**
+     * Generates a simple text-only {@link SGraph} for a Graph with only the given labels.
+     * The {@code label}s are expected to come from the {@code modelLabels} field after the {@link #toSGraph}
+     * translation.
+     * 
+     * @param labels A list of all labels, this SGraph should contain.
+     * @param parentId The ID of the graph containing all these labels.
+     */
+    static def SGraph generateTextDiagram(List<SKLabel> labels, String parentId) {
+        // equivalent for the SRootElement
+        val root = new SKGraph => [
+            type = 'graph'
+            id = parentId + KGraphElementIdGenerator.ID_SEPARATOR + "texts-only"
+            children = new ArrayList
+        ]
+
+        root.children = new ArrayList
+        root.children += labels
+        return root
+    }
+    
+    def restoreState(KGraphDiagramState state) {
+        throw new UnsupportedOperationException("TODO: auto-generated method stub")
+    }
+    
+}
```

Comment:
I did not look through all of this class, as this is a lot of duplicated code from the non-incremental diagram generator. To avoid that they should get in some relation to each other (inheritance, shared methods or similar).